### DISCUSSION TITLE
tccutil and SIP

### DISCUSSION THREAD

**Discussion by jacobsalmela** [jacobsalmela/tccutil#44] [Q&A]
Since I currently don't plan to circumvent Apple's SIP, I'm porting over the discussions from these two issues to a Discussion.

- #18 
- #40 
- https://github.com/jacobsalmela/tccutil/issues/62

So please continue the discussion here.

**Comment by chwzr** [comment]
do we have a proof that tccutil will work ( read / write) with SIP disabled as of now (big sur)?

**Reply by jacobsalmela** [reply]
Did you see? https://github.com/jacobsalmela/tccutil/pull/41

It's in https://github.com/jacobsalmela/tccutil/releases/tag/v1.2.10

**Reply by jeffest-jeff** [reply]
@jacobsalmela , just to clarify real quick (I apologize if this has already been answered)...basically, just disable SIP when you are using tccutil to modify the database. After that, SIP can be re-enabled?
If so, this could be used during an initial deployment.

**Reply by mkfl3x** [reply]
@jeff89179, as I know you can [turn on/off](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection) SIP any time you want with recovery mode.
@chwzr, in Big Sur disabled SIP allows to use tccutil (at least for me)

**Reply by jacobsalmela** [reply]
@jeff89179 yeah, that's what I meant.  It should work with SIP disabled.  I don't have any plans to circumvent SIP, so that's the solution for now.

**Reply by jeffest-jeff** [reply]
@jacobsalmela that's completely fine. Just wanted to clarify. Thanks!
@mkfl3x thanks!

**Reply by ylluminate** [reply]
With SIP entirely off on 11.2.3 I'm getting:
```
$ sudo /usr/local/bin/tccutil --enable <tld.bundle.name>
TCC Database structure is unknown.
Error opening Database.  You probably need to disable SIP for this to work.
```

**Reply by jacoblb64** [reply]
FWIW I'm getting the same error on 11.4, SIP is already disabled

**Reply by ylluminate** [reply]
Can this be used to implement support for Big Sur?: https://rainforest.engineering/2021-02-09-macos-tcc/

**Reply by ylluminate** [reply]
Managed to get something to work from the shell. In this example I'm enabling Zoom.us.app:
```
# get bundle id:
$ osascript -e 'id of app "zoom.us"'

# use it in the following commands:
$ sudo sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db "INSERT or REPLACE INTO access VALUES('kTCCServiceMicrophone','us.zoom.xos',0,2,0,1,NULL,NULL,NULL,'UNUSED',NULL,0,1541440109);"

$ sudo sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db "INSERT or REPLACE INTO access VALUES('kTCCServiceCamera','us.zoom.xos',0,2,0,1,NULL,NULL,NULL,'UNUSED',NULL,0,1541440109);"

```

**Reply by jacobsalmela** [reply]
@ylluminate can you check out https://github.com/jacobsalmela/tccutil/commit/9f3251c4819afe44662ae937d6c44a4a7112f0ab and then run `./tccutil.py --digest` to see what it returns?

**Reply by ylluminate** [reply]
@jacobsalmela,
```
$ git clone ...
$ git checkout bigsur
$ sudo ./tccutil.py --digest
3d1c2a0e97
```

**Reply by ylluminate** [reply]
@jacobsalmela can you please clarify your confused emoticon reply?

**Reply by jacobsalmela** [reply]
Oh, I was trying to track down that digest hash, which seems to change each release of macOS.  I was expecting something different, so I'm not sure if it's my code or how to track down the right digest just yet.

**Reply by ylluminarious** [reply]
@jacobsalmela I'm also experiencing a weird issue here. I was working with @ylluminate on another Big Sur machine and had this problem happen on it.

I tried running:

```
sudo ./tccutil.py --enable com.bundle.id
```

And the output was:
```
Error opening Database.  You probably need to disable SIP for this to work.
```

(I already have SIP disabled)

I also tried running: 

```
sudo ./tccutil.py --digest
```

And there was no output. The exit status was also `0`:

```
$ echo $?                   
0
```

(i.e., it wasn't an error)

**Reply by jacobsalmela** [reply]
Ok.  I need to go back and check the code then.  This would be a useful addition to the code base I think.  

Let me ask in one of the PR threads how to get the hash and see if I'm on the right track or wandering far from the beacon

**Reply by jacobsalmela** [reply]
Can anyone here following this thread, check out this branch: https://github.com/jacobsalmela/tccutil/compare/bigsur

and see what the `--digest` flag returns?

**Reply by ylluminate** [reply]
@jacobsalmela tried to checkout that branch but it now appears to be gone...?

**Reply by jacobsalmela** [reply]
Merged to main

**Reply by ylluminate** [reply]
Oh, well then on 11.6:
```
$ sudo ./tccutil.py --digest
3d1c2a0e97
```

**Comment by ylluminate** [comment]
Any news or developments on this front? Or is anyone aware of a project that may have taken up the torch on this front? 

It seems that there are ongoing issues and I'm curious about how to proceed when we do things like use `amfi_get_out_of_my_way` and no longer have an easy method to manage TCC.

Also, @jacobsalmela, have you thought about rewriting this project in V? Crossed my mind a bit here as it would be interesting, but I'm also very low on time.

**Reply by jacobsalmela** [reply]
I have thought about a rewrite in go or v.  

In terms of SIP, I haven't planned to circumvent it.  I am a bit rusty on it's inner workings the past few releases as well.  

What specifically were you wondering about?  This app working with/around SIP?

**Reply by ylluminate** [reply]
Well, specifically I've been wanting to kinda facilitate more easily operating with as many of macOS "features" turned off as possible. I have so many more strange bugs and weird things happening with their updates and it's easy to see that most of the points of pain are from both "safety" and convenience features that simply aren't necessary. I'm not the kind of person that feels as though things should be forced upon you if you're even at least generally cognizant of the consequences of turning them off and you shouldn't be punished for doing so. To this end, being able to still use devices and tcc-facilitated functions should not be eliminated if you choose to operate with `amfi_get_out_of_my_way` on, etc. 

It seems that this app simply isn't working again on 12.6.4 and I've been kinda thinking outloud about what I can do to improve the situation. I still cannot "upgrade" to Ventura given some of the serious problems I had when I had initially tried to migrate some months ago and hesitate to try again until perhaps... well I dunno. I just want my macOS-as-a-Linux-option days back.

**Comment by danilagalimov** [comment]
Actually, there is a bug in newest version which incorrectly reports the error, hiding the actual one: `TypeError: Unicode-objects must be encoded before hashing`.

Suggested change:
line 110 -> accessTableDigest = hashlib.sha1(row[0]**.encode('utf-8')**).hexdigest()[0:10]
after line 157 add `print(traceback.format_exc())` to print exception verbosely.

and add `import traceback` in the beginning.

**Reply by jacobsalmela** [reply]
@danilagalimov want to make a PR?  Much easier to see and test in that way.

**Comment by danilagalimov** [comment]
Actually, it looks like it is already fixed here, but brew package installs 1.2.11 version instead of latest

**Reply by jacobsalmela** [reply]
https://github.com/Homebrew/homebrew-core/pull/140587

**Comment by danilagalimov** [comment]
It works for me to read the settings, but for writing it seems that schema has changed - now it has the hash in **csreq** column and 'auth_value' should be 2

**Comment by danilagalimov** [comment]
Googled an example how to calculate csreq column- https://stackoverflow.com/questions/52706542/how-to-get-csreq-of-macos-application-on-command-line
